File: backend/scraper.js
const fs = require('fs');

const departments = {
    'architecture': 'Architecture',
    'cecase': 'CECASE',
    'chem': 'Chemical Engineering',
    'chemistry': 'Chemistry',
    'civil': 'Civil Engineering',
    'ca': 'Computer Applications',
    'cse': 'Computer Science & Engineering',
    'dee': 'DEE',
    'eee': 'Electrical & Electronics Engineering',
    'ece': 'Electronics & Communication Engineering',
    'humanities': 'Humanities',
    'ice': 'Instrumentation & control Engineering',
    'management': 'Management Studies',
    'maths': 'Mathematics',
    'mech': 'Mechanical Engineering',
    'meta': 'Metallurgical & Materials Engineering',
    'physics': 'Physics',
    'prod': 'Production Engineering'
};

const seed_file_header = `import { PrismaClient } from '@prisma/client'
import bcrypt from 'bcryptjs'

const prisma = new PrismaClient()

async function main() {
  const hashedPassword = await bcrypt.hash('password123', 10)

  // Super Admin
  await prisma.user.upsert({
    where: { email: 'admin@nitt.edu' },
    update: {},
    create: {
      email: 'admin@nitt.edu',
      password: hashedPassword,
      role: 'SUPER_ADMIN',
    },
  })
`;

const seed_file_footer = `
  console.log('Seed completed.')
}

main()
  .catch((e) => {
    console.error(e)
    process.exit(1)
  })
  .finally(async () => {
    await prisma.$disconnect()
  })
`;

async function scrape() {
    let output = seed_file_header;

    for (const [code, name] of Object.entries(departments)) {
        console.log(`Fetching ${name} (${code})...`);
        output += `\n  // ${name} Department\n`;
        output += `  const dept_${code} = await prisma.department.upsert({\n`;
        output += `    where: { code: '${code}' },\n`;
        output += `    update: {},\n`;
        output += `    create: { name: '${name}', code: '${code}' },\n`;
        output += `  })\n\n`;

        try {
            const res = await fetch(`https://www.nitt.edu/home/academics/departments/${code}/faculty/`, {
                headers: {
                    'User-Agent': 'Mozilla/5.0'
                }
            });
            const text = await res.text();
            
            // Simple regex to extract <a ...>Dr. ...</a> or <a ...>Mr. ...</a>
            const regex = /<a[^>]*>(Dr\.[^<]+|Mr\.[^<]+|Ms\.[^<]+|Prof\.[^<]+)<\/a>/g;
            let match;
            const faculties = [];

            while ((match = regex.exec(text)) !== null) {
                const name = match[1].trim().replace(/\s+/g, ' ');
                if (!faculties.includes(name) && name.length > 4) {
                    faculties.push(name);
                }
            }

            // Fallback if regex fails (some pages might format names differently)
            if (faculties.length === 0) {
                const altRegex = /<a[^>]*>([^<]+)<\/a>/g;
                while ((match = altRegex.exec(text)) !== null) {
                    const altName = match[1].trim().replace(/\s+/g, ' ');
                    if ((altName.startsWith('Dr.') || altName.startsWith('Mr.') || altName.startsWith('Ms.')) && !faculties.includes(altName)) {
                        faculties.push(altName);
                    }
                }
            }

            if (faculties.length === 0) {
                faculties.push(`Dr. Dummy ${code.toUpperCase()}`);
            }

            faculties.forEach((facultyName, i) => {
                let emailBase = facultyName.split(' ').pop().toLowerCase().replace(/[^a-z0-9]/g, '').substring(0, 5);
                const email = `${emailBase}${i}_${code}@nitt.edu`;
                const safeName = facultyName.replace(/'/g, "\\'");
                const role = i === 0 ? 'HOD' : 'FACULTY';
                
                output += `  await prisma.user.upsert({\n`;
                output += `    where: { email: '${email}' },\n`;
                output += `    update: {},\n`;
                output += `    create: {\n`;
                output += `      email: '${email}',\n`;
                output += `      password: hashedPassword,\n`;
                output += `      role: '${role}',\n`;
                output += `      facultyProfile: {\n`;
                output += `        create: {\n`;
                output += `          departmentId: dept_${code}.id,\n`;
                output += `          name: '${safeName}',\n`;
                output += `          designation: '${role === 'HOD' ? 'Head of Department' : 'Professor'}',\n`;
                output += `          researchTags: 'General, Engineering',\n`;
                output += `          profileCompletion: 50\n`;
                output += `        }\n`;
                output += `      }\n`;
                output += `    },\n`;
                output += `  })\n\n`;
            });

        } catch (e) {
            console.error(`Error fetching ${code}: ${e.message}`);
        }
    }

    output += seed_file_footer;
    fs.writeFileSync('prisma/seed.ts', output);
    console.log('Successfully generated prisma/seed.ts');
}

scrape();

File: backend/scraper.py
import requests
from bs4 import BeautifulSoup
import re

departments = {
    'architecture': 'Architecture',
    'cecase': 'CECASE',
    'chem': 'Chemical Engineering',
    'chemistry': 'Chemistry',
    'civil': 'Civil Engineering',
    'ca': 'Computer Applications',
    'cse': 'Computer Science & Engineering',
    'dee': 'DEE',
    'eee': 'Electrical & Electronics Engineering',
    'ece': 'Electronics & Communication Engineering',
    'humanities': 'Humanities',
    'ice': 'Instrumentation & control Engineering',
    'management': 'Management Studies',
    'maths': 'Mathematics',
    'mech': 'Mechanical Engineering',
    'meta': 'Metallurgical & Materials Engineering',
    'physics': 'Physics',
    'prod': 'Production Engineering'
}

seed_file_header = """const { PrismaClient } = require('@prisma/client')
const bcrypt = require('bcryptjs')

const prisma = new PrismaClient()

async function main() {
  const hashedPassword = await bcrypt.hash('password123', 10)

  // Super Admin
  await prisma.user.upsert({
    where: { email: 'admin@nitt.edu' },
    update: {},
    create: {
      email: 'admin@nitt.edu',
      password: hashedPassword,
      role: 'SUPER_ADMIN',
    },
  })

"""

seed_file_footer = """
  console.log('Seed completed.')
}

main()
  .catch((e) => {
    console.error(e)
    process.exit(1)
  })
  .finally(async () => {
    await prisma.$disconnect()
  })
"""

with open('prisma/seed.ts', 'w') as f:
    f.write(seed_file_header)

    for code, name in departments.items():
        print(f"Fetching {name} ({code})...")
        f.write(f"\n  // {name} Department\n")
        f.write(f"  const dept_{code} = await prisma.department.upsert({{\n")
        f.write(f"    where: {{ code: '{code}' }},\n")
        f.write(f"    update: {{}},\n")
        f.write(f"    create: {{ name: '{name}', code: '{code}' }},\n")
        f.write(f"  }})\n\n")

        try:
            url = f"https://www.nitt.edu/home/academics/departments/{code}/faculty/"
            res = requests.get(url, timeout=10)
            soup = BeautifulSoup(res.text, 'html.parser')

            faculties = []
            
            # Find all links in the main content area that look like faculty names
            content_div = soup.find('div', id='content') or soup.body
            links = content_div.find_all('a')
            
            for link in links:
                text = link.text.strip()
                # Check if it starts with Dr. or contains regular faculty name format
                if text.startswith('Dr') or text.startswith('Mr') or text.startswith('Ms') or text.startswith('Prof'):
                    if text not in faculties and len(text) > 4:
                        faculties.append(text)

            # If no faculties found, just add a dummy
            if not faculties:
                faculties = [f"Dr. Dummy {code.upper()}"]

            for i, faculty_name in enumerate(faculties):
                email = f"{faculty_name.split()[-1].lower()[:5]}{i}_{code}@nitt.edu"
                email = re.sub(r'[^a-zA-Z0-9@.]', '', email)
                safe_name = faculty_name.replace("'", "\\'")
                
                # Determine role - first faculty is HOD
                role = "HOD" if i == 0 else "FACULTY"
                
                f.write(f"  await prisma.user.upsert({{\n")
                f.write(f"    where: {{ email: '{email}' }},\n")
                f.write(f"    update: {{}},\n")
                f.write(f"    create: {{\n")
                f.write(f"      email: '{email}',\n")
                f.write(f"      password: hashedPassword,\n")
                f.write(f"      role: '{role}',\n")
                f.write(f"      facultyProfile: {{\n")
                f.write(f"        create: {{\n")
                f.write(f"          departmentId: dept_{code}.id,\n")
                f.write(f"          name: '{safe_name}',\n")
                f.write(f"          designation: '{'Head of Department' if role == 'HOD' else 'Professor'}',\n")
                f.write(f"          researchTags: 'General, Engineering',\n")
                f.write(f"          profileCompletion: 50\n")
                f.write(f"        }}\n")
                f.write(f"      }}\n")
                f.write(f"    }},\n")
                f.write(f"  }})\n\n")

        except Exception as e:
            print(f"Error fetching {code}: {e}")

    f.write(seed_file_footer)
